src/main/java/org/wildfly/security/auth/server/CredentialMapper.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2015 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.wildfly.security.auth.server;
+
+import org.wildfly.security.sasl.util.SaslMechanismInformation.Names;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Credential selection mapper mechanism consume authentication process information and use it to yield a credential name.
+ * Provided to ServerAuthenticationContext for determine credential name(s) which should be acquired from security realm.
+ *
+ * @author <a href="mailto:jkalina@redhat.com">Jan Kalina</a>
+ */
+public interface CredentialMapper {
+
+    /**
+     * Get credential names by authentication information.
+     * @param information the authentication information (at least mechanism type, name and user name)
+     * @return the list of credential names
+     */
+    List<String> getCredentialNameMapping(AuthenticationInformation information);
+
+    CredentialMapper ELYTRON_CREDENTIAL_MAPPER = information -> {
+        String mech = information.getMechanismName();
+        if (Names.DIGEST_MD5.equals(mech)) return Arrays.asList("userPassword-md5","userPassword-clear","clear");
```

Comment:
When defining these names I think they need to be in terms of strings that an authentication mechanism could be expected to use when interacting with the SecurityRealm API - once we start talking about LDAP attributes it has really moved onto an implementation detail of the realm itself.  If a user swaps an LDAP realm for a JDBC realm the mechanisms wouldn't be expected to query based on column name.